modules/renter/repair.go

```diff
@@ -26,323 +25,374 @@ var renewThreshold = func() types.BlockHeight {
+			chunkPrefix := make([]byte, 8)
+			binary.LittleEndian.PutUint64(chunkPrefix, i)
+			chunkName := string(chunkPrefix) + file.name
 
-	incomplete := make(map[uint64][]uint64)
-	for chunkIndex, pieceBools := range present {
-		for pieceIndex, ok := range pieceBools {
-			if !ok {
-				incomplete[uint64(chunkIndex)] = append(incomplete[uint64(chunkIndex)], uint64(pieceIndex))
-			}
+			// Add the chunk to the repair matrix.
+			repairMatrix[chunkName] = &gaps
 		}
 	}
-	return incomplete
 }
 
-// chunkHosts returns the hosts storing the given chunk.
-func (f *file) chunkHosts(chunk uint64) []modules.NetAddress {
-	f.mu.RLock()
-	defer f.mu.RUnlock()
+func (r *Renter) createRepairMatrix(availableWorkers map[types.FileContractID]struct{}) (map[string]*chunkGaps, map[int]int) {
+	repairMatrix := make(map[string]*chunkGaps)
+	gapCounts := make(map[int]int)
 
-	var old []modules.NetAddress
-	for _, fc := range f.contracts {
-		for _, p := range fc.Pieces {
-			if p.Chunk == chunk {
-				old = append(old, fc.IP)
-				break
-			}
-		}
+	// Add all of the files to the repair matrix.
+	for _, file := range r.files {
+		r.addFileToRepairMatrix(file, availableWorkers, repairMatrix, gapCounts)
 	}
-	return old
+	return repairMatrix, gapCounts
 }
 
-// expiringContracts returns the contracts that will expire soon.
-// TODO: what if contract has fully expired?
-func (f *file) expiringContracts(height types.BlockHeight) []fileContract {
-	f.mu.RLock()
-	defer f.mu.RUnlock()
-
-	var expiring []fileContract
-	for _, fc := range f.contracts {
-		if height >= fc.WindowStart-renewThreshold {
-			expiring = append(expiring, fc)
+func (r *Renter) managedRepairIteration() {
+	// Create the initial set of workers that are used to perform
+	// uploading.
+	availableWorkers := make(map[types.FileContractID]struct{})
+	id := r.mu.RLock()
+	for id, worker := range r.workerPool {
+		// Ignore workers that have had an upload failure in the past two
+		// hours.
+		if worker.recentUploadFailure.Add(time.Minute * 120).Before(time.Now()) {
+			availableWorkers[id] = struct{}{}
 		}
 	}
-	return expiring
-}
+	r.mu.RUnlock(id)
 
-// offlineChunks returns the chunks belonging to "offline" hosts -- hosts that
-// do not meet uptime requirements. Importantly, only chunks missing more than
-// half their redundancy are returned.
-func (f *file) offlineChunks(hdb hostDB) map[uint64][]uint64 {
-	f.mu.RLock()
-	defer f.mu.RUnlock()
-
-	// mark all pieces belonging to offline hosts.
-	offline := make(map[uint64][]uint64)
-	for _, fc := range f.contracts {
-		if hdb.IsOffline(fc.IP) {
-			for _, p := range fc.Pieces {
-				offline[p.Chunk] = append(offline[p.Chunk], p.Piece)
-			}
+	// Create the repair matrix. The repair matrix is a set of chunks,
+	// linked from chunk id to the set of hosts that do not have that
+	// chunk.
+	repairMatrix, gapCounts := r.createRepairMatrix(availableWorkers)
+	maxGaps := 0
+	for i := 1; i < len(gapCounts); i++ {
+		if gapCounts[i] > 0 {
+			maxGaps = i
```

Comment:
I'm confused by `gapCounts`. What is it a `map[int]int` if you don't ever use the second `int`? Could it be a `[]bool`? Could the "maximum" code be handled inside `createRepairMatrix`, so that it just returns `maxGaps`?